transcom/mymove
# INT B-20273 add and update indexes for move history

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20273)

## Summary

We are having some performance issues in staging that is related to this query. The AWS cloudwatch logs show this error:
```
@message {"level":"error","ts":"2024-05-15T21:34:57.631Z","caller":"ghcapi/move_history.go:40",
"msg":"Error retrieving move history by locator",
"git_branch":"main",
"git_commit":"6b21b53bfd0754d38b66bee9a5dd85b35566c58d",
"ecs_cluster":"arn:aws-us-gov:ecs:us-gov-west-1:015932076428:cluster/app-stg",
"ecs_task_def_family":"app-stg",
"ecs_task_def_revision":"4785",
"host":"office.stg.move.mil",
"milmove_trace_id":"66452a77-3703-4b99-3d8e-700d19eb4b2f",
"session_id":"9PiZPmdfcn9gAchp-o6fKwFLvJnilpbSwE9gViPtsLc",
"xray_trace_id":"1-66452a77-37034b993d8e700d19eb4b2f",
"error":"Could not complete query related to object of type: AuditHistory.",
```

Which we found to be erroring out in `pkg/services/move_history/move_history_fetcher.go:49`:
```
	rawQuery, qerr := query.GetSQLQueryByName("move_history_fetcher")
...
	query := appCtx.DB().RawQuery(rawQuery, locator).Paginate(int(*params.Page), int(*params.PerPage))
	err := query.All(audits)

	if err != nil {
		switch err {
		case sql.ErrNoRows:
			// Not found error expects an id but we're querying by locator
			return &models.MoveHistory{}, 0, apperror.NewNotFoundError(uuid.Nil, "move locator "+locator)
		default:
			return &models.MoveHistory{}, 0, apperror.NewQueryError("AuditHistory", err, "")
		}
	}
```

So the issue is with the query in `move_history_fetcher.sql`, which is basically a big fat query that pulls data from all of the important tables (`moves`, `mto_shipments`, `orders`, `mto_service_items`, etc).

Unfortunately we can't simulate the performance issues locally because staging has a ridiculous amount of data in their RDS table and locally we tend to clean a lot and have a lot less data. 

The first round of solutions to improve the performance issues besides refactoring is making sure we have indexes set up in all of the necessary places. Below are all of the important tables and the columns that are used in several `JOIN` and subqueries within the CTEs in `move_history_fetcher.sql`:

`moves`
`id`: Used in multiple subqueries and joins.
`locator`: Used in the initial filtering condition in the move CTE.

`mto_shipments`
`id`: Used in joins with `audit_history` and `move_service_items`
`move_id`: Used in the filtering condition in the `move_shipments` CTE
`destination_address_id`, `secondary_delivery_address_id`, `pickup_address_id`, `secondary_pickup_address_id`: Used in the `move_addresses` CTE

`audit_history`
`object_id`: Used in multiple joins
`table_name`: Used in the filtering conditions for various logs
`event_name` and `changed_data`: Used in filtering conditions in several places

`orders`
`id`: Used in joins and subqueries.
`service_member_id`: Used in the `move_service_members` CTE

`mto_service_items`
`id`: Used in joins with `audit_history`.
`move_id`: Used in the filtering condition in the `move_service_items` CTE
`mto_shipment_id`: Used in joins with `move_shipments`

`re_services`
`id`: Used in joins with `move_service_items`

`entitlements`
`id`: Used in joins with `audit_history`

`payment_requests`
`id`: Used in joins with `audit_history`.
`move_id`: Used in filtering condition in the `move_payment_requests` CTE

`payment_service_items`
`payment_request_id`: Used in joins with `payment_requests`

`proof_of_service_docs`
`payment_request_id`: Used in joins with `move_payment_requests`

`mto_agents`
`id`: Used in joins with `audit_history`
`mto_shipment_id`: Used in joins with `move_shipments`

`reweighs`
`id`: Used in joins with `audit_history`.
`shipment_id`: Used in joins with `move_shipments`

`service_members`
`id`: Used in joins and subqueries
`residential_address_id`, `backup_mailing_address_id`: Used in the `move_addresses` CTE

`sit_address_updates`
`id`: Used in joins with `audit_history`
`mto_service_item_id`: Used in joins with `move_service_items`

`user_uploads`
`id`: Used in joins with `audit_history`
`document_id`: Used in joins with `documents`

`documents`
`id`: Used in joins with `user_uploads`.
`service_member_id`: Used in filtering condition in `file_uploads` CTE

`weight_tickets`
`ppm_shipment_id`: Used in joins with `ppm_shipments`

`progear_weight_tickets`
`ppm_shipment_id`: Used in joins with `ppm_shipments`

`moving_expenses`
`ppm_shipment_id`: Used in joins with `ppm_shipments`

`ppm_shipments`
`shipment_id`: Used in joins with `move_shipments`

`backup_contacts`
`service_member_id`: Used in filtering condition in `move_backup_contacts` CTE

`addresses`
`id`: Used in joins with `audit_history`

SOOOOOO all that being said, the only indexes that did not exist already were:
`audit_history.object_id`
`proof_of_service_docs.payment_request_id`
`mto_agents.mto_shipment_id`
`ppm_shipments.shipment_id`
`backup_contacts.service_member_id`

This PR adds those indexes. Sorry for the long read. Thanks for hangin'.

### How to test

1. Open up MM as an office user
2. Review the move history logs and make sure that you can still see them without errors
3. Review DBeaver and make sure you see the new indexes on the above tables and their respective columns (find table, click "Properties", then click "Indexes").
